runtime/cudaq/distributed/builtin/activate_custom_mpi.sh

```diff
@@ -21,27 +21,35 @@
 #                ${MPI_PATH}/include is expected to contain the mpi.h header.
 #                ${MPI_PATH}/lib64 or ${MPI_PATH}/lib is expected to contain libmpi.so.
 #
-# Run: source <cuda quantum install dir>/distributed_interfaces/activate_custom_mpi.sh
+# Run: bash <cuda quantum install dir>/distributed_interfaces/activate_custom_mpi.sh
 # 
 # You could add $CUDAQ_MPI_COMM_LIB to your ~/.bashrc file to persist the environment variable.
 
 if [ -z "${MPI_PATH}" ]
 then
     echo "Environment variable MPI_PATH is not set. Please set it to point to the MPI root directory!"
     echo "Note that MPI_PATH/include is expected to contain the mpi.h header."
-    return
+    (return 0 2>/dev/null) && return 1 || exit 1
```

Comment:
Is this entire file sourced into a function someplace? Not sure what the `return` from `function` is for here since this doesn't appear to be a `function` context nor why there are two `return` statements. Why not just use `exit 1`?

We guess is `return` isn't in a bash `function` so is generating an error. That error is discarded by sending it to `/dev/null`. The second `return` is never executed because the first one is an error. Finally, the `exit 1` is executed.

Reply:
when the script is sourced (i.e. executed with `source activate_custom_mpi.sh` or `. activate_custom_mpi.sh`), then return is the correct way to terminate the script. If the script uses exit, then that would make the calling shell exit, and hence sourcing scripts that contain exit when sourced is not save. This setup ensures that independent on how the script is called, it will always terminate in a way that the command has the correct exit code and never makes the calling shell quit.

How the line works: If the script is not sourced (i.e run with `bash activate_custom_mpi.sh` or `./activate_custom_mpi.sh`), the first return 0 command runs without terminating the script but producing a non-zero code for the command, such that the exit 1 is triggered, which then terminates the script with a code 1. If the script is sourced, then the second return is triggered ending the script with code 1 - this may seem odd, but you can confirm this by trying it out with the script
```
(return 0 2>/dev/null) && echo after check && return 1 || exit 1
echo "should never get here"
```
The best explanation I have is that &&, ||, if-statements etc are special; e.g. also with regards to exiting when a command fails;
```
bheim@BHEIM-DT1:~/makeself$ set -e
bheim@BHEIM-DT1:~/makeself$ bash -c "exit 1" || true
bheim@BHEIM-DT1:~/makeself$ bash -c "exit 1"
[process exited with code 1 (0x00000001)]
```